chore(android): gradle 8

AbdullahFaqeir opened the pull request:
chore(android): let there be gradle 8
chore(android): fixed deprecations in kroll-apt project
chore(android): made kroll-apt incremental to enhance build time

**GitHub issue:** https://github.com/tidev/titanium-sdk/issues/13579

This PR is dedicated to my mentor Nafe Abboushi.

AbdullahFaqeir committed:
chore(android): let there be gradle 8
chore(android): fixed deprecations in kroll-apt project
chore(android): made kroll-apt incremental to enhance build time

m1ga reviewed (changes requested):
Are the changes to the JAVA files necessary for the update? The more stuff that is changed, the more we have to test :wink: 

Also I wouldn't update the min and targetSDK with this PR. If we still can use 21 we should do that too. Same with target 34. I have a PR for that here https://github.com/tidev/titanium-sdk/pull/13940/ as the broadcast receiver has to be adjusted.

So if that is not needed for the gradle PR please keep it as it is and do it in a separate PR (less changes, less testing).

I couldn't build a SDK locally. It's almost working but the last step `titaniumPublication(MavenPublication)` tries to publish it online instead of making a local repo? I've canceled it as it was uploading stuff :smile: Didn't want to publish something I don't need.

There aren't all cmake version available during build process. I have a different PR about that https://github.com/tidev/titanium-sdk/pull/13966 and I had to use `3.22.1` as that is available on all platforms.

AbdullahFaqeir committed:
fix(android): revert change to java files
fix(android): gradle 8 compatibility update to template build.gradle
fix(android): revert dependency updates
fix(android): revert minSdk and targetSdk

AbdullahFaqeir reviewed (commented):
All reviews applied.

AbdullahFaqeir committed:
fix(android): kitchensink-app not running after gradle 8
fix(android): update _build.js & _buildModule.js for new gradle

m1ga commented:
I have to do some testing but the first round was very good and I was able to build kitchensink an my app now :+1: 

Review will follow once I have more time to test

cb1kenobi commented on android/app/build.gradle:
```diff
@@ -9,47 +9,50 @@ apply plugin: 'com.android.application'
 			assets.srcDirs = [
-				'src/main/assets',
-				"${projectDir}/../titanium/build/outputs/ti-assets"
+					'src/main/assets',
+					"${projectDir}/../titanium/build/outputs/ti-assets"
```
Are these indentations intentional?

```suggestion
				'src/main/assets',
				"${projectDir}/../titanium/build/outputs/ti-assets"
```

cb1kenobi commented on android/templates/build/ti.constants.gradle:
```diff
@@ -6,12 +6,13 @@
 	tiPlayServicesBaseLibVersion = '18.2.0'
 	tiManifestPlaceholders = [
-		tiActivityConfigChanges: 'density|fontScale|keyboard|keyboardHidden|layoutDirection|locale|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode'
+			tiActivityConfigChanges: 'density|fontScale|keyboard|keyboardHidden|layoutDirection|locale|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode'
```
Intentional indention?

```suggestion
		tiActivityConfigChanges: 'density|fontScale|keyboard|keyboardHidden|layoutDirection|locale|mcc|mnc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|touchscreen|uiMode'
```

cb1kenobi commented on android/titanium/build.gradle:
```diff
@@ -57,9 +59,9 @@ android {
-					'kroll.outputJsonFilePath': "${projectDir}/../../dist/android/titanium.bindings.json".toString(),
-					'kroll.outputCppDirPath': "${projectDir}/../runtime/v8/generated".toString(),
-					'kroll.jsModuleName': 'titanium'
+						'kroll.outputJsonFilePath': "${projectDir}/../../dist/android/titanium.bindings.json".toString(),
+						'kroll.outputCppDirPath'  : "${projectDir}/../runtime/v8/generated".toString(),
+						'kroll.jsModuleName'      : 'titanium'
```
Intentional indention?

```suggestion
					'kroll.outputJsonFilePath': "${projectDir}/../../dist/android/titanium.bindings.json".toString(),
					'kroll.outputCppDirPath'  : "${projectDir}/../runtime/v8/generated".toString(),
					'kroll.jsModuleName'      : 'titanium'
```

cb1kenobi commented on android/titanium/build.gradle:
```diff
@@ -82,50 +87,45 @@ android {
 			assets.srcDirs = [
-				'assets',
-				"${projectDir}/../modules/ui/assets"
+					'assets',
+					"${projectDir}/../modules/ui/assets"
```
Intentional indention?

```suggestion
				'assets',
				"${projectDir}/../modules/ui/assets"
```

cb1kenobi commented on android/titanium/build.gradle:
```diff
@@ -82,50 +87,45 @@ android {
 			res.srcDirs = [
-				'res',
-				"${projectDir}/../modules/ui/res"
+					'res',
+					"${projectDir}/../modules/ui/res"
```
Intentional indention?

```suggestion
				'res',
				"${projectDir}/../modules/ui/res"
```

cb1kenobi commented on android/titanium/build.gradle:
```diff
@@ -139,8 +139,8 @@ task updateV8Library() {
 		v8MakeFile.text = [
-			"set(LIBV8_VERSION \"${packageJson.v8.version}\")",
-			"set(LIBV8_MODE \"${packageJson.v8.mode}\")"
+				"set(LIBV8_VERSION \"${packageJson.v8.version}\")",
+				"set(LIBV8_MODE \"${packageJson.v8.mode}\")"
```
Intentional indention?

```suggestion
			"set(LIBV8_VERSION \"${packageJson.v8.version}\")",
			"set(LIBV8_MODE \"${packageJson.v8.mode}\")"
```

cb1kenobi commented on android/untar.gradle:
```diff
@@ -27,10 +27,10 @@ task untar() {
-		overwrite: project.properties.overwrite,
-		src: project.properties.src,
-		dest: project.properties.dest)
+			compression: project.properties.compression,
+			overwrite: project.properties.overwrite,
+			src: project.properties.src,
+			dest: project.properties.dest)
```
Intentional indention?

```suggestion
		compression: project.properties.compression,
		overwrite: project.properties.overwrite,
		src: project.properties.src,
		dest: project.properties.dest)
```

cb1kenobi reviewed (commented):
This is AWESOME! I don't wanna lose momentum on this PR.

I found a few indention nitpicks. I haven't tested the PR yet, but I spent some time researching some of the changes such as no longer need to call `.setPackageName()`. Good stuff!

What sort of testing do we need to do to get this over the finish line? I can test on Windows doing emulator builds as well as build-only device/dist builds. I don't have any working Android devices. Sigh.

m1ga commented:
> What sort of testing do we need to do to get this over the finish line?

:+1:  @cb1kenobi I would love to split this up into smaller steps as this is a big PR:
https://github.com/tidev/titanium-sdk/issues/13579#issuecomment-2021150043
I have made older PRs: first one only updates the gradle files to the newer syntax, 2nd one updates cmake and checkstyle (looks much but it's only linting). Then we only have the gradle 8 parts in here and can focus on that as it's a bigger change.

hansemannn commented on android/app/build.gradle:
```diff
@@ -9,47 +9,50 @@ apply plugin: 'com.android.application'
 		applicationId 'com.titanium.test'
-		minSdkVersion 21
+		compileSdk 33
+		minSdkVersion 26
```
Doesn't this bump the default min-sdk version to Android 8? We should remain stable with the current one

hansemannn commented on android/app/build.gradle:
```diff
@@ -9,47 +9,50 @@ apply plugin: 'com.android.application'
 android {
-	compileSdkVersion 33
+	ndkVersion project.ext.tiNdkVersion
+	//namespace 'com.titanium.test'
```
Remove comment

hansemannn committed:
Update android/app/build.gradle

Co-authored-by: Chris Barber <chris@cb1inc.com>

hansemannn commented on android/build.gradle:
```diff
@@ -6,17 +6,17 @@
 	dependencies {
-		classpath 'com.android.tools.build:gradle:7.0.4'
-		classpath 'com.google.gms:google-services:4.3.15'
+		classpath 'com.android.tools.build:gradle:8.3.1'
```
Let's do 8.10.1

hansemannn commented on android/cli/commands/_buildModule.js:
```diff
@@ -612,12 +612,6 @@ AndroidModuleBuilder.prototype.generateModuleProject = async function generateMo
-		packageName = `ti.${packageName}`;
-	}
-	mainManifest.setPackageName(packageName);
-	await mainManifest.writeToFilePath(path.join(moduleMainDir, 'AndroidManifest.xml'));
```
What happened to this?

hansemannn commented on android/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,7 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.4-bin.zip
```
Use latest one as well

hansemannn commented on android/templates/build/root.build.gradle:
```diff
@@ -1,14 +1,13 @@
-		classpath 'com.android.tools.build:gradle:7.0.4'
-		classpath 'com.google.gms:google-services:4.3.10'
+		classpath 'com.android.tools.build:gradle:8.3.1'
+		classpath 'com.google.gms:google-services:4.4.1'
```
Use latest ones for kotlin, gradle and google-services

hansemannn committed:
Update android/templates/build/ti.constants.gradle

Co-authored-by: Chris Barber <chris@cb1inc.com>

hansemannn committed:
Update android/titanium/build.gradle

Co-authored-by: Chris Barber <chris@cb1inc.com>

hansemannn committed:
Update android/titanium/build.gradle

Co-authored-by: Chris Barber <chris@cb1inc.com>

hansemannn committed:
Update android/untar.gradle

Co-authored-by: Chris Barber <chris@cb1inc.com>

hansemannn reviewed (changes requested):
A few questions and changes we should verify. And can you please rebase with latest master to resolve the merge conflicts?

AbdullahFaqeir committed:
Update android/titanium/build.gradle

Co-authored-by: Chris Barber <chris@cb1inc.com>

AbdullahFaqeir commented:
Was this put in the system in any way or not, I hope so, and if not I would love to get back and work on it

AbdullahFaqeir commented:
> A few questions and changes we should verify. And can you please rebase with latest master to resolve the merge conflicts?

Of course

AbdullahFaqeir committed:
Merge branch 'master' into gradle-upgrade

m1ga commented on android/app/build.gradle:
```diff
@@ -9,7 +9,8 @@ apply plugin: 'com.android.application'
 
 // Set up Android app project.
 android {
-	compileSdkVersion 33
```
this is still needed. It complains that it is missing. It can be 34

m1ga commented:
Did a first quick test. The current PR won't build. These files need to be changed:

android/modules/ui/src/java/ti/modules/titanium/ui/PickerProxy.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiImageView.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIActivityIndicator.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButton.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButtonBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIOptionBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUITabbedBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIText.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiRecyclerViewHolder.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/searchbar/TiUISearchBar.java
android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIAbstractTabGroup.java

 Some `R.` resources need to be changed. E.g.
```
R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox_ExposedDropdownMenu
```
will be
```
com.google.android.material.R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox_ExposedDropdownMenu
```

We can't change the import `org.appcelerator.titanium.R` because some lines like `R.style.Widget_Titanium_OutlinedButton_IconOnly` are in the titanium namespace.

Best is to follow the errors when building the SDK (~30 changes). I'm not sure how to add files to this PR :blush:  But it won't take long to change it. Just go through the error messages and change `R` with `com.google.android.material.R`.

After that: SDK builds fine and I tested it with two apps with existing modules and they work fine. Kitchensink works too. Hyperloop-examples breaks: 
```
[ERROR] [GRADLE] A problem occurred configuring project ':gradle-project'.
[ERROR] [GRADLE] > Could not create an instance of type com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
[ERROR] [GRADLE]    > Namespace not specified. Specify a namespace in the module's build file. See https://d.android.com/r/tools/upgrade-assistant/set-namespace for information about setting the namespace.
[ERROR] [GRADLE]      
[ERROR] [GRADLE]      If you've specified the package attribute in the source AndroidManifest.xml, you can use the AGP Upgrade Assistant to migrate to the namespace value in the build file. Refer to https://d.android.com/r/tools/upgrade-assistant/agp-upgrade-assistant for general information about using the AGP Upgrade Assistant.
[ERROR] [GRADLE] 
```

m1ga commented on android/gradle.properties:
```diff
@@ -14,4 +14,6 @@ android.useAndroidX=true
 android.enableJetifier=false
 
 # remove if com.android.tools.build:gradle can be raised in /android/build.gradle
+
 android.suppressUnsupportedCompileSdk=33,34
+
```
This can be removed now

hansemannn commented:
@m1ga PR's can be checked out most easily with the Github CLI and `gh pr checkout 14014` in this case. It clones the branch and checks it out. You can then push into the branch and the commits occur here.

m1ga committed:
fix material.r

m1ga commented:
@hansemannn nice! That worked :+1: SDK will build now, modules & hyperloop currently fail with the error above. 

@AbdullahFaqeir would be nice if you can find out where the namespace is missing. Apps without hyperloop work fine but I didn't check all controls yet.

m1ga committed:
github action

hansemannn commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIButton.java:
```diff
@@ -60,23 +59,23 @@ public TiUIButton(final TiViewProxy proxy)
 		switch (styleId) {
 			case UIModule.BUTTON_STYLE_OPTION_POSITIVE:
-				styleId = R.attr.buttonBarPositiveButtonStyle;
+				styleId = com.google.android.material.R.attr.buttonBarPositiveButtonStyle;
```
To keep the diff small, you could also just `import com.google.android.material.R;` I think.

m1ga committed:
optimize material imports

m1ga committed:
optimize material imports

AbdullahFaqeir committed:
fix(android): fix module building

AbdullahFaqeir commented:
@m1ga @hansemannn I fixed it to build for modules, couldn't build tho on my Silicon machine, but it the fix for sure went after that stage.

m1ga commented:
awesome! SDK builds, apps build, modules build :+1: We are getting closer.

* fix for hyperloop apps: https://github.com/tidev/hyperloop.next/pull/392
* small workflow update in https://github.com/tidev/titanium-sdk/pull/14112 (modules build fine, but when you start a example app build in a module it won't forward the `--sdk 12.6.0` parameter)

m1ga committed:
Merge branch 'master' into gradle-upgrade

m1ga commented:
one thing we'll need to tell the module developers: minSdk 12.6.0 is required when you build modules with this SDK as it is building with a higher Java version. The other way (old modules in 12.6.0 apps) works fine.

AbdullahFaqeir commented:
> one thing we'll need to tell the module developers: minSdk 12.6.0 is required when you build modules with this SDK as it is building with a higher Java version. The other way (old modules in 12.6.0 apps) works fine.

@m1ga while generating the manifest file?

m1ga commented:
when building an app with a module that was build with this PR you'll see
´Unsupported class file major version 61` as the module was build with JAVA_17 and the app is trying to build with JAVA_11 (current 12.4.0). But since we have the `minSdk` value in the module manifest it's fine. The developer just have to update it when using 12.6.0 to build the module

AbdullahFaqeir committed:
Merge branch 'master' into gradle-upgrade

AbdullahFaqeir commented:
> when building an app with a module that was build with this PR you'll see ´Unsupported class file major version 61`as the module was build with JAVA_17 and the app is trying to build with JAVA_11 (current 12.4.0). But since we have the`minSdk` value in the module manifest it's fine. The developer just have to update it when using 12.6.0 to build the module

@m1ga Exactly!

AbdullahFaqeir commented:
@m1ga  @hansemannn I believe the PR is ready to be merged?

m1ga committed:
Merge branch 'master' into gradle-upgrade

m1ga reviewed (commented):
My tests look good:
* building the SDK :heavy_check_mark: 
* building app with multiple existing modules with this SDK :heavy_check_mark: 
* building modules and the included example app :heavy_check_mark: 
* building hyperloop-examples :heavy_check_mark: 
* building kitchensink-v2 :heavy_check_mark:

hansemannn commented on android/templates/module/generated/build.gradle:
```diff
@@ -18,7 +18,7 @@ android {
 	compileSdkVersion <%- compileSdkVersion %>
 	ndkVersion '22.1.7171670'
```
@AbdullahFaqeir You resolved this but added no commit?

AbdullahFaqeir replied:
Ooh, didn't see yours, I believe yes it should, and it allows me to build on my machine, let me see this

AbdullahFaqeir replied:
Done

AbdullahFaqeir committed:
fix(android): bump ndk version for generated modules

AbdullahFaqeir committed:
Merge branch 'gradle-upgrade' of https://github.com/AbdullahFaqeir/titanium-sdk into gradle-upgrade

hansemannn reviewed (approved):
LGTM!

AbdullahFaqeir commented:
LET'S GOOOO 🚀🚀🚀

cb1kenobi commented:
Excellent work!

AhmedMSayed commented:
Halla walla, That's Awesome @AbdullahFaqeir you saved my day
